src/RelayProvider.ts

```diff
@@ -0,0 +1,270 @@
+import { constants, Transaction } from 'ethers';
+import type { Networkish } from "@ethersproject/networks";
+import { JsonRpcProvider, TransactionReceipt } from "@ethersproject/providers";
+import type { ConnectionInfo } from "@ethersproject/web";
+import { getEnvelopingConfig } from './utils';
+import { hexValue } from "@ethersproject/bytes";
+import RelayClient, { RequestConfig } from './RelayClient';
+import log from 'loglevel';
+import { RelayHub__factory } from '@rsksmart/rif-relay-contracts';
+import AccountManager from './AccountManager';
+import type { UserDefinedEnvelopingRequest } from './common/relayRequest.types';
+
+export interface RelayingResult {
+    validUntilTime?: string;
+    transaction: Transaction;
+    receipt: TransactionReceipt;
+}
+
+export default class RelayProvider extends JsonRpcProvider {
+
+    //TODO: remove & when relayTransaction PR is merged and rebased
+    private readonly relayClient!: RelayClient & { relayTransaction: (
+        envelopingRequest: UserDefinedEnvelopingRequest,
+        requestConfig: RequestConfig) => Promise<Transaction> };
+    
+    private readonly jsonRpcProvider: JsonRpcProvider;
+
+    constructor(url?: ConnectionInfo | string, network?: Networkish, relayClient?: RelayClient){
+        super(url, network);
+        this.jsonRpcProvider = new JsonRpcProvider(url, network);
+        //TODO: remove this when relayTransaction PR is merged and rebased
+        //@ts-ignore
+        this.relayClient =
+            relayClient ??
+            new RelayClient();
+    }
+
+    private _useEnveloping(method: string, requestConfig: RequestConfig): boolean {
+        if (method === 'eth_accounts') {
+            return true;
+        }
+        if (!requestConfig) {
+            return false;
+        }
+
+        return requestConfig.useEnveloping ?? false;
+    }
+
+    private _getRelayStatus(respResult: TransactionReceipt): {
+        relayRevertedOnRecipient: boolean;
+        transactionRelayed: boolean;
+        reason?: string;
+    } {
+        if (!respResult.logs || respResult.logs.length === 0) {
+            return {
+                relayRevertedOnRecipient: false,
+                transactionRelayed: false,
+                reason: 'Tx logs not found'
+            };
+        }
+
+        const relayHubInterface = RelayHub__factory.createInterface();
+        const parsedLogs = respResult.logs.map(log => relayHubInterface.parseLog(log));
```

Comment:
This type is not exported on the contracts repo, I would have to import it by going all the way into the dist folder and I don't think this is desirable.

Reply:
It is desirable to define the type, otherwise the variable can represent anything. The having to go to dist should have been mitigated, but I seem to remember that it didn't work or something, so it's just something we'll need to resolve sooner or later, def not a reason to not use the generated types from typechain.

Reply:
I dont understand, which one is the one that does not have the typing? We have an index in the contracts where we can get or export any desired type from typechain.

Reply:
`import type { RelayHubInterface } from '@rsksmart/rif-relay-contracts/dist/typechain-types/contracts/RelayHub';
`

Reply:
I can add it in the rif-relay-contracts. 

```
export {
  EnvelopingTypes,
  StakeUnlockedEvent,
  IForwarder as IForwarderTypes,
  RelayHubInterface
} from './typechain-types/contracts/RelayHub';

Reply:
@ironFe93 I think this is still valid, isn't it?